File: webhooks/nodejs/samples/folder.copied.js
// Rename Copied Folder

const express = require("express");
const crypto = require("crypto");
const axios = require("axios");
require("dotenv").config();

const app = express();
app.use(express.json({ verify: rawBodySaver }));

// Save raw body for signature verification
function rawBodySaver(req, res, buf) {
  req.rawBody = buf.toString();
}

// Generate HMAC-SHA256 signature
function getSecretHash(secret, payload) {
  return (
    "sha256=" +
    crypto.createHmac("sha256", secret).update(payload).digest("hex")
  );
}

app.post("/webhook", async (req, res) => {
  const signature = req.headers["x-docspace-signature-256"];
  const computedSignature = getSecretHash(
    process.env.WEBHOOK_SECRET_KEY,
    req.rawBody
  );

  if (signature !== computedSignature) {
    console.warn("Invalid signature");
    return res.sendStatus(401);
  }

  const { trigger } = req.body.event;
  const {
    id: folderId,
    title: folderTitle,
    createBy
  } = req.body.payload;

  // Handle folder copied event
  if (trigger === "folder.copied") {
    console.log(`Folder "${folderTitle}" was copied (ID: ${folderId})`);

    try {
      // Generate new folder name with date
      const today = new Date().toISOString().slice(0, 10); // Format: YYYY-MM-DD
      const newTitle = `${folderTitle}_COPY_${today}`;

      // Rename the folder via API
      await axios.put(
        `${process.env.PORTAL_URL}/api/2.0/folders/${folderId}/rename`,
        { title: newTitle },
        {
          headers: {
            Authorization: `Bearer ${process.env.API_KEY}`,
            "Content-Type": "application/json"
          }
        }
      );

      console.log(`Folder renamed to "${newTitle}"`);

      // Optionally: notify the user who copied the folder
      const userInfo = await axios.get(
        `${process.env.PORTAL_URL}/api/2.0/people/${createBy}`,
        {
          headers: {
            Authorization: `Bearer ${process.env.API_KEY}`,
            "Content-Type": "application/json"
          }
        }
      );

      const email = userInfo.data?.email;
      if (email) {
        await axios.post(
          process.env.NOTIFY_WEBHOOK_URL,
          {
            to: email,
            subject: `Folder copied`,
            message: `You copied the folder "${folderTitle}". It has been renamed to "${newTitle}".`
          },
          { headers: { "Content-Type": "application/json" } }
        );

        console.log(`Notification sent to ${email}`);
      }

      return res.sendStatus(200);
    } catch (err) {
      console.error("Error handling folder.copied:", err.response?.data || err.message);
      return res.sendStatus(500);
    }
  }

  res.sendStatus(200);
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Webhook server listening on port ${PORT}`);
});

File: webhooks/python/samples/folder.copied.py
# Rename Copied Folder

from flask import Flask, request, abort
import hmac
import hashlib
import requests
import os
from datetime import datetime

app = Flask(__name__)

# Generate HMAC-SHA256 signature
def get_secret_hash(secret, payload):
    return "sha256=" + hmac.new(secret.encode(), payload, hashlib.sha256).hexdigest()

@app.route("/webhook", methods=["POST"])
def webhook():
    signature = request.headers.get("x-docspace-signature-256")
    computed_signature = get_secret_hash(
        os.environ.get("WEBHOOK_SECRET_KEY"),
        request.get_data()
    )

    if signature != computed_signature:
        app.logger.warning("Invalid signature")
        abort(401)

    event = request.json.get("event", {})
    payload = request.json.get("payload", {})

    trigger = event.get("trigger")
    folder_id = payload.get("id")
    folder_title = payload.get("title")
    created_by = payload.get("createBy")

    # Handle folder copied event
    if trigger == "folder.copied":
        app.logger.info(f'Folder "{folder_title}" was copied (ID: {folder_id})')

        try:
            # Generate new folder name with date
            today = datetime.utcnow().strftime("%Y-%m-%d")  # Format: YYYY-MM-DD
            new_title = f"{folder_title}_COPY_{today}"

            # Rename the folder via API
            requests.put(
                f"{os.environ.get('PORTAL_URL')}/api/2.0/folders/{folder_id}/rename",
                json={"title": new_title},
                headers={
                    "Authorization": f"Bearer {os.environ.get('API_KEY')}",
                    "Content-Type": "application/json"
                }
            )

            app.logger.info(f'Folder renamed to "{new_title}"')

            # Optionally: notify the user who copied the folder
            user_info = requests.get(
                f"{os.environ.get('PORTAL_URL')}/api/2.0/people/{created_by}",
                headers={
                    "Authorization": f"Bearer {os.environ.get('API_KEY')}",
                    "Content-Type": "application/json"
                }
            ).json()

            email = user_info.get("email")
            if email:
                requests.post(
                    os.environ.get("NOTIFY_WEBHOOK_URL"),
                    json={
                        "to": email,
                        "subject": "Folder copied",
                        "message": f'You copied the folder "{folder_title}". It has been renamed to "{new_title}".'
                    },
                    headers={"Content-Type": "application/json"}
                )

                app.logger.info(f"Notification sent to {email}")

            return "", 200

        except Exception as err:
            app.logger.error(f"Error handling folder.copied: {err}")
            return "", 500

    return "", 200

if __name__ == "__main__":
    port = int(os.environ.get("PORT", 3000))
    app.run(host="0.0.0.0", port=port)
